### ISSUE TITLE
tf.train.batch does not preserve the order of data and miss some data.

### ISSUE THREAD

**Issue by HC-2016** [tensorflow/tensorflow#12384]
I use tf.train.batch to produce data in tensorflow but get unexpected data order as following: 
```
# Create a queue that produces the filenames to read.
filename_queue = tf.train.string_input_producer(string_tensor=filenames, num_epochs=1, shuffle=False)

......

num_preprocess_threads = 1 
images_x, images_y, label_batch = tf.train.batch(
            [image[0], image[1], label],
            batch_size=1,
            num_threads=num_preprocess_threads,
            capacity=1)
```

the label produced by tf.train.batch: [ 6.03125     5.2734375   4.03125   5.84375     5.15625   ....]. A total of 33 labels.

the label of original data: [ 6.03125     5.2734375   **4.546875**    4.03125     **6.1875**  5.8438 ...]. A total of 68 labels.

It seems that the generated data differs from the original data in the order and is less. Why?

**Comment by qinglintian** [comment]
seems that I met the same problem

**Comment by HC-2016** [comment]
@qinglintian Any solutions?

**Comment by qinglintian** [comment]
Nope. Actually, I met this problem yesterday and haven't got time to explore further.

**Comment by HC-2016** [comment]
@qinglintian 
When I delete some specific variables in the parameter fetch of sess.run(), it works as expected. It seems that the computing graph has been evaluated more than one times in a round.

**Comment by mrry** [comment]
@HC-2016 Can you share a complete reproduction for the problem?  Based on your last comment, it sounds like you are fetching tensors from both "before" and "after" the `tf.train.batch()` in the same `sess.run()` call. This will lead to some input values being consumed before they get to be batched. Unfortunately this kind of error is hard for TensorFlow to detect, since the `sess.run()`  call doesn't know anything about the queue-runner structure.

(As an aside, the `tf.contrib.data` API might be useful in preventing this kind of bug. It also has better facilities for getting a deterministic execution order, even when shuffling is used.)

**Comment by HC-2016** [comment]
@mrry sorry, I can not share the codes for some reason now. And I think you are right. I have fetched tensors from both "before" and "after" the tf.train.batch() in the same sess.run() call. Stopping to fetch the tensor before tf.train.batch() indeed solves the problem.
